### PR TITLE
Feature: Purge local room history.

### PR DESCRIPTION
**Pull request by erikjohnston**

### CODE REVIEW

File: synapse/storage/events.py [hunk 1]
```diff
@@ -1281,6 +1281,151 @@ def get_all_new_events_txn(txn):
...
+            "ON e.event_id = f.event_id "
+            "AND e.room_id = f.room_id "
+            "WHERE f.room_id = ?",
+            (room_id,)
+        )
+        rows = txn.fetchall()
+        max_depth = max(row[0] for row in rows)
+
+        if max_depth <= topological_ordering:
+            raise Exception("topological_ordering is greater than forward extremeties")
```
Comment (NegativeMjark) [inline synapse/storage/events.py]: Maybe 400 here rather than 500?
Maybe comment on why this is bad.

**Comment by NegativeMjark** [comment]
What happens when clients try to backfill with tokens from before the purge?

**Comment by erikjohnston** [comment]
> What happens when clients try to backfill using tokens from before the purge?

It (successfully) returns 0 events. It probably triggers backfill to remote servers, so eventually it would probably return some events once the sever had backpaginated enough.

**Comment by NegativeMjark** [comment]
LGTM

**Comment by rubo77** [comment]
Is this already a solution for https://github.com/matrix-org/synapse/issues/890 ?

**Comment by erikjohnston** [comment]
@rubo77 Yes, though its only a basic, manual per room API atm

**Comment by 4nd3r** [comment]
@erikjohnston is this manual API documented somewhere or can anyone can give an example?

**Comment by rubo77** [comment]
In  #916 I posted a gist, that could work. If it works, we should add that in the api-rst file

**Comment by erikjohnston** [comment]
@4nd3r It's briefly documented at: https://github.com/matrix-org/synapse/blob/master/docs/admin_api/purge_history_api.rst

**Comment by rubo77** [comment]
@erikjohnston can you take part of my example gist from #916 into that documentation?

https://gist.github.com/rubo77/458eb4808d665e619feeaf556476f73f

**Comment by 4nd3r** [comment]
@erikjohnston thanks. does this API also delete uploaded media from disk?

**Comment by rubo77** [comment]
I updated to 0.17 now, and now my gist file works fine

**Comment by rubo77** [comment]
https://github.com/matrix-org/synapse/pull/1034 will add an example to the API Documentation

**Pull request merged**